shared/calc.go

```diff
@@ -0,0 +1,28 @@
+package shared
+
+import (
+	"math/big"
+)
+
+// Deviates checks if x1 and x2 deviates based on the provided ppb (parts per billion)
+// ppb is calculated based on the smaller value of the two
+// e.g, if x1 > x2, deviation_parts_per_billion = ((x1 - x2) / x2) * 1e9
+func Deviates(x1, x2 *big.Int, ppb int64) bool {
+	// if x1 == 0 or x2 == 0, deviates if x2 != x1, to avoid the relative division by 0 error
+	if x1.BitLen() == 0 || x2.BitLen() == 0 {
+		return x1.Cmp(x2) != 0
+	}
+	diff := big.NewInt(0).Sub(x1, x2) // diff = x1-x2
+	diff.Mul(diff, big.NewInt(1e9))   // diff = diff * 1e9
+	// dividing by the smaller value gives consistent ppb regardless of input order, and supports >100% deviation.
+	if x1.Cmp(x2) > 0 {
+		diff.Div(diff, x2)
+	} else {
+		diff.Div(diff, x1)
+	}
+	return diff.CmpAbs(big.NewInt(ppb)) > 0 // abs(diff) > ppb
```

Comment:
Still not sure what it does, but maybe this is slightly clearer? Still not sure what it means by >100% deviation.
```suggestion
	// ensure x1 > x2
	if x1.Cmp(x2) < 0 {
		x1, x2 = x2, x1
	}
	diff := big.NewInt(0).Sub(x1, x2) // diff = x1-x2
	diff.Mul(diff, big.NewInt(1e9))   // diff = diff * 1e9
	// dividing by the smaller value gives consistent ppb regardless of input order, and supports >100% deviation.
	diff.Div(diff, x2)
	return diff.Cmp(big.NewInt(ppb)) > 0 // diff > ppb
```